File: backend/app/models/ingest.py
"""
Provenance for everything that enters a case.

One row per file the ingestion pipeline has seen, whatever door it came
through. This is the single answer to "what has been ingested into this case,
where did it come from, and what did we decide it was?" — a question the
fourteen independent upload endpoints could not answer between them.

Written with SQLAlchemy 2.0's `Mapped` annotations rather than the bare
`Column()` the older models use. That is not a style preference: `Column()`
leaves mypy seeing `Column[str]` where the code assigns a `str`, and the
type-checking ratchet in `pyproject.toml` forbids adding new modules to the
exemption list. Annotated columns type-check as what they actually hold.

See `docs/INGESTION.md` sections 2 and 4.
"""
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import BigInteger, ForeignKey, Index, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from ..database import Base


def _uuid() -> str:
    return str(uuid.uuid4())


# ─── States ───────────────────────────────────────────────────────────────────
# The pipeline is a state machine, not a boolean. `docs/INGESTION.md` section 4
# holds the diagram; the design rule it encodes is that **no state discards the
# file**. Unrecognised input is a soft signal, never an error.

STATE_DISCOVERED   = "discovered"      # seen, stable, not yet read
STATE_HASHED       = "hashed"          # SHA-256 computed
STATE_DUPLICATE    = "duplicate"       # hash already in this case. Terminal.
STATE_IDENTIFIED   = "identified"      # type determined
STATE_UNIDENTIFIED = "unidentified"    # no signature matched. Recoverable.
STATE_ROUTED       = "routed"          # destination chosen, record created
STATE_PARSED       = "parsed"          # converted to tabular form
STATE_INDEXED      = "indexed"         # queryable in the Artifact Explorer
# Stored, and a module opens it as it stands - no table, and none wanted. A
# registry hive is the case this exists for: `SOFTWARE` holds thousands of
# unrelated facts and which of them matter is an analyst's decision, so Remora
# supplies the navigation instead of the conclusions. Reporting that as
# `unsupported` said the pipeline had no answer when it had a better one than a
# table would have been. Terminal.
STATE_BROWSABLE    = "browsable"
STATE_FAILED       = "failed"          # parser crashed / timed out. Recoverable.
STATE_UNSUPPORTED  = "unsupported"     # known type, no handler yet

#: States a file will never leave without an analyst acting on it.
TERMINAL_STATES = frozenset({
    STATE_DUPLICATE, STATE_INDEXED, STATE_BROWSABLE, STATE_UNSUPPORTED,
})

#: States the Collection tab offers an action for: force a type, retry.
RECOVERABLE_STATES = frozenset({STATE_UNIDENTIFIED, STATE_FAILED})

ALL_STATES = frozenset({
    STATE_DISCOVERED, STATE_HASHED, STATE_DUPLICATE, STATE_IDENTIFIED,
    STATE_UNIDENTIFIED, STATE_ROUTED, STATE_PARSED, STATE_INDEXED,
    STATE_BROWSABLE, STATE_FAILED, STATE_UNSUPPORTED,
})

# ─── Origins ──────────────────────────────────────────────────────────────────
# How the bytes reached the drop folder. Every value lands in the same folder
# and goes through the same service; this records only which door was used.

ORIGIN_DROPZONE  = "dropzone"   # copied onto the host: cp, scp, rsync, a mount
ORIGIN_UPLOAD    = "upload"     # the Collection tab's drop area
ORIGIN_ARCHIVE   = "archive"    # unpacked from a container already ingested
ORIGIN_CONNECTOR = "connector"  # pulled by an integration
# Ingested before this table existed, by one of the fourteen upload endpoints.
# Which door it used was never recorded, and inventing an answer would be
# fabricating provenance in a tool whose output is meant to survive scrutiny.
# "We do not know" is the only honest value, so it is a value.
ORIGIN_LEGACY    = "legacy"

ALL_ORIGINS = frozenset({
    ORIGIN_DROPZONE, ORIGIN_UPLOAD, ORIGIN_ARCHIVE, ORIGIN_CONNECTOR,
    ORIGIN_LEGACY,
})

# ─── Detection sources ────────────────────────────────────────────────────────
# Recorded because they are not equally trustworthy. A file identified by
# `extension` is a guess an analyst may need to override; one identified by
# `magic` is not.

DETECTED_BY_MAGIC     = "magic"        # byte signature
DETECTED_BY_EXTENSION = "extension"    # filename suffix, no signature matched
DETECTED_BY_HINT      = "folder_hint"  # the evtx/ or registry/ sub-folder
DETECTED_BY_CONTENT   = "content"      # text-shape heuristic (CSV vs JSON vs log)
DETECTED_BY_FORCED    = "forced"       # an analyst said so. Outranks everything.


class IngestedFile(Base):
    """One file seen by the ingestion pipeline."""

    __tablename__ = "ingested_files"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=_uuid)

    # No index=True: `ix_ingested_files_case_sha` leads on case_id and SQLite
    # uses its leftmost prefix for a plain case lookup. A third index on the
    # same leading column would only cost writes.
    case_id: Mapped[str] = mapped_column(
        String, ForeignKey("cases.id", ondelete="CASCADE"), nullable=False)

    # Groups a batch, or an archive and everything unpacked out of it, into one
    # logical import. Nullable while the legacy Collection Import still owns
    # some ingests (`docs/INGESTION.md` section 14, step 1).
    collection_id: Mapped[str | None] = mapped_column(
        String, ForeignKey("imported_collections.id", ondelete="CASCADE"),
        nullable=True, index=True)

    # An archive member points at the container it came out of, so a whole
    # import traces back to the one file the analyst actually dropped.
    parent_id: Mapped[str | None] = mapped_column(
        String, ForeignKey("ingested_files.id", ondelete="CASCADE"),
        nullable=True, index=True)

    original_name: Mapped[str]         = mapped_column(String, nullable=False)
    stored_path:   Mapped[str | None]  = mapped_column(String, nullable=True)
    size_bytes:    Mapped[int]         = mapped_column(BigInteger, default=0)

    origin:        Mapped[str]         = mapped_column(
        String, nullable=False, default=ORIGIN_DROPZONE)
    origin_detail: Mapped[str | None]  = mapped_column(String, nullable=True)

    # Indexed on its own as well as with the case: deduplication is per case,
    # but "where else has this exact file been seen?" is worth being able to
    # answer across an installation, and the composite cannot serve it.
    sha256: Mapped[str | None] = mapped_column(String, nullable=True, index=True)

    magic_type:       Mapped[str | None] = mapped_column(String, nullable=True)
    detected_kind:    Mapped[str | None] = mapped_column(String, nullable=True)
    detection_source: Mapped[str | None] = mapped_column(String, nullable=True)

    # IANA name. UTC is a default, not a fact — hence the warning the Collection
    # tab shows when a collection is left at the default and its parser produced
    # naive timestamps (`docs/INGESTION.md` section 8).
    source_timezone: Mapped[str] = mapped_column(String, default="UTC")

    # Likewise covered by `ix_ingested_files_case_state`. Alone, state has ten
    # possible values and no planner would choose an index over a scan.
    state: Mapped[str]        = mapped_column(
        String, nullable=False, default=STATE_DISCOVERED)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)

    routed_to:          Mapped[str | None] = mapped_column(String, nullable=True)
    parsed_artifact_id: Mapped[str | None] = mapped_column(String, nullable=True)

    # Set once the file has been preserved in the chain of custody. Its
    # presence is what exempts the file from the 90-day collection expiry:
    # a preserved copy lives in the evidence store, which nothing expires.
    # The constraint is named: SQLite's batch migration cannot drop an unnamed
    # one, so an anonymous foreign key here makes the downgrade unrunnable.
    evidence_id: Mapped[str | None] = mapped_column(
        String,
        ForeignKey("evidences.id", ondelete="SET NULL",
                   name="fk_ingested_files_evidence_id"),
        nullable=True, index=True)

    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow, onupdate=datetime.utcnow)

    __table_args__ = (
        # Deduplication is per case: the same EVTX legitimately belongs to two
        # different investigations, and blocking it globally would corrupt the
        # second one. Not unique — a duplicate is recorded rather than refused,
        # so the analyst can see that they dropped it twice.
        Index("ix_ingested_files_case_sha", "case_id", "sha256"),
        Index("ix_ingested_files_case_state", "case_id", "state"),
    )

File: frontend/src/api/ingest.ts
/**
 * The ingest queue.
 *
 * What the pipeline has seen for a case and what it decided, plus the two
 * actions the recoverable states call for. This is the view that makes
 * "what has been ingested here?" answerable - the question fourteen
 * independent upload endpoints could not answer between them.
 */
const BASE = '/api/v1'

export type IngestState =
  | 'discovered' | 'hashed' | 'duplicate' | 'identified' | 'unidentified'
  | 'routed' | 'parsed' | 'indexed'
  /** Stored, and a module opens it as it stands. A registry hive. Terminal. */
  | 'browsable'
  | 'failed' | 'unsupported'

export interface IngestedFile {
  id:                string
  original_name:     string
  size_bytes:        number
  origin:            'dropzone' | 'upload' | 'archive' | 'connector' | 'legacy'
  origin_detail:     string | null
  sha256:            string | null
  magic_type:        string | null
  detected_kind:     string | null
  detection_source:  'magic' | 'extension' | 'folder_hint' | 'content' | 'forced' | null
  source_timezone:   string | null
  state:             IngestState
  error:             string | null
  routed_to:         string | null
  parent_id:         string | null
  collection_id:     string | null
  /** The analyst can act on this one: force a type, or retry it. */
  recoverable:       boolean
  evidence_id:       string | null
  /** Preserved in the chain of custody, so the collection expiry no longer applies. */
  preserved:         boolean
  destination_pages: string[]
  created_at:        string | null
}

export interface IngestKind {
  kind:        string
  destination: string | null
  parser:      string | null
  /** False when the type is recognised but its parser has not shipped. */
  available:   boolean
}

function authHeaders(): Record<string, string> {
  const token = localStorage.getItem('remora_token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

async function unwrap<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const body = await response.json().catch(() => null)
    throw new Error(body?.detail ?? `Request failed (${response.status})`)
  }
  return response.status === 204 ? (undefined as T) : response.json()
}

export const ingestApi = {
  list(caseId: string, state?: IngestState) {
    const query = state ? `?state=${encodeURIComponent(state)}` : ''
    return fetch(`${BASE}/cases/${caseId}/ingest${query}`, { headers: authHeaders() })
      .then(r => unwrap<{ files: IngestedFile[]; summary: Record<string, number> }>(r))
  },

  kinds() {
    return fetch(`${BASE}/ingest/kinds`, { headers: authHeaders() })
      .then(r => unwrap<{ kinds: IngestKind[] }>(r))
  },

  /** Override the detected type. Outranks every other detection source. */
  forceKind(caseId: string, fileId: string, kind: string) {
    return fetch(`${BASE}/cases/${caseId}/ingest/${fileId}/force-kind`, {
      method:  'POST',
      headers: { ...authHeaders(), 'Content-Type': 'application/json' },
      body:    JSON.stringify({ kind }),
    }).then(r => unwrap<IngestedFile>(r))
  },

  /**
   * Tell the pipeline which OS a raw memory image came from, and parse it.
   *
   * Only raw dumps need this. A Windows crash dump and a LiME image say which
   * OS they are in their header, and are parsed without asking.
   */
  setMemoryOs(caseId: string, fileId: string, osType: 'windows' | 'linux') {
    return fetch(`${BASE}/cases/${caseId}/ingest/${fileId}/memory-os`, {
      method:  'POST',
      headers: { ...authHeaders(), 'Content-Type': 'application/json' },
      body:    JSON.stringify({ os_type: osType }),
    }).then(r => unwrap<IngestedFile>(r))
  },

  retry(caseId: string, fileId: string) {
    return fetch(`${BASE}/cases/${caseId}/ingest/${fileId}/retry`, {
      method: 'POST', headers: authHeaders(),
    }).then(r => unwrap<IngestedFile>(r))
  },

  /** Where to copy files by hand, for the analyst not going through the browser. */
  dropPath(caseId: string) {
    return fetch(`${BASE}/cases/${caseId}/ingest/drop-path`, { headers: authHeaders() })
      .then(r => unwrap<{ container_path: string; folder_name: string }>(r))
  },
}
